File: server/nomadflow/routers/features.py
"""Feature (worktree) management endpoints."""

from pathlib import Path

from fastapi import APIRouter, Depends, HTTPException

from ..config import Settings, get_settings
from ..auth import verify_auth
from ..services.git_service import GitService
from ..services.tmux_service import TmuxService
from ..models.requests import (
    ListFeaturesRequest,
    CreateFeatureRequest,
    DeleteFeatureRequest,
    SwitchFeatureRequest,
)
from ..models.responses import (
    ListFeaturesResponse,
    CreateFeatureResponse,
    DeleteFeatureResponse,
    SwitchFeatureResponse,
)

router = APIRouter(prefix="/api", tags=["features"], dependencies=[Depends(verify_auth)])


def get_git_service(settings: Settings = Depends(get_settings)) -> GitService:
    """Dependency to get GitService instance."""
    return GitService(settings)


def get_tmux_service(settings: Settings = Depends(get_settings)) -> TmuxService:
    """Dependency to get TmuxService instance."""
    return TmuxService(settings)


def get_window_name(repo_path: str, feature_name: str) -> str:
    """Build tmux window name from repo and feature.

    Format: repo:feature (e.g., 'my-project:add-login')
    This prevents collisions when different repos have features with the same name.
    """
    repo_name = Path(repo_path).name
    return f"{repo_name}:{feature_name}"


@router.post("/list-features", response_model=ListFeaturesResponse)
async def list_features(
    request: ListFeaturesRequest,
    git_service: GitService = Depends(get_git_service),
) -> ListFeaturesResponse:
    """
    List all features (worktrees) for a repository.

    Returns a list of features with their names, worktree paths, branches,
    and active status.
    """
    try:
        features = await git_service.list_features(request.repo_path)
        return ListFeaturesResponse(features=features)
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))


@router.post("/create-feature", response_model=CreateFeatureResponse)
async def create_feature(
    request: CreateFeatureRequest,
    git_service: GitService = Depends(get_git_service),
    tmux_service: TmuxService = Depends(get_tmux_service),
) -> CreateFeatureResponse:
    """
    Create a new feature worktree.

    Creates a git worktree for the feature and sets up a tmux window
    for working on it.
    """
    try:
        # Create the git worktree (auto-detects base branch if not specified or "main")
        base_branch = request.base_branch if request.base_branch != "main" else None
        worktree_path, branch = await git_service.create_feature(
            request.repo_path,
            request.feature_name,
            base_branch,
        )

        # Ensure tmux session exists
        await tmux_service.ensure_session()

        # Create or ensure tmux window for this feature
        # Window name format: repo:feature to avoid collisions
        window_name = get_window_name(request.repo_path, request.feature_name)
        await tmux_service.ensure_window(window_name, worktree_path)

        return CreateFeatureResponse(
            worktree_path=worktree_path,
            branch=branch,
            tmux_window=window_name,
        )
    except RuntimeError as e:
        raise HTTPException(status_code=400, detail=str(e))
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))


@router.post("/delete-feature", response_model=DeleteFeatureResponse)
async def delete_feature(
    request: DeleteFeatureRequest,
    git_service: GitService = Depends(get_git_service),
    tmux_service: TmuxService = Depends(get_tmux_service),
) -> DeleteFeatureResponse:
    """
    Delete a feature worktree.

    Removes the git worktree and optionally the associated branch.
    Also kills any associated tmux window.
    """
    try:
        # Kill tmux window if it exists
        # Window name format: repo:feature
        window_name = get_window_name(request.repo_path, request.feature_name)
        await tmux_service.kill_window(window_name)

        # Delete the git worktree
        deleted = await git_service.delete_feature(
            request.repo_path,
            request.feature_name,
        )

        return DeleteFeatureResponse(deleted=deleted)
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))


@router.post("/switch-feature", response_model=SwitchFeatureResponse)
async def switch_feature(
    request: SwitchFeatureRequest,
    git_service: GitService = Depends(get_git_service),
    tmux_service: TmuxService = Depends(get_tmux_service),
) -> SwitchFeatureResponse:
    """
    Switch to a feature worktree.

    This endpoint:
    1. Finds or creates the feature's worktree
    2. Ensures the tmux session exists
    3. Creates/selects the tmux window for the feature
    4. CDs into the worktree directory
    5. Clears the terminal

    After calling this endpoint, the ttyd terminal will show the active
    tmux window positioned in the correct worktree directory.
    """
    try:
        # Get features to find the worktree path
        features = await git_service.list_features(request.repo_path)
        feature = next(
            (f for f in features if f.name == request.feature_name),
            None,
        )

        worktree_path: str

        if not feature:
            # Feature doesn't exist, create it automatically
            # create_feature auto-detects the default branch
            worktree_path, branch = await git_service.create_feature(
                request.repo_path,
                request.feature_name,
                None,  # Auto-detect base branch
            )
        else:
            worktree_path = feature.worktree_path

        # Ensure tmux session exists
        await tmux_service.ensure_session()

        # Switch to window (creates if needed, selects, cds, and clears)
        # Returns (switched, has_running_process) - won't send cd/clear if process is running
        # Window name format: repo:feature to avoid collisions
        window_name = get_window_name(request.repo_path, request.feature_name)
        switched, has_running_process = await tmux_service.switch_to_window(
            window_name,
            worktree_path,
        )

        if not switched:
            raise HTTPException(
                status_code=500,
                detail=f"Failed to switch to window '{window_name}'",
            )

        return SwitchFeatureResponse(
            switched=True,
            worktree_path=worktree_path,
            tmux_window=window_name,
            has_running_process=has_running_process,
        )
    except RuntimeError as e:
        raise HTTPException(status_code=400, detail=str(e))
    except HTTPException:
        raise
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))

File: nomadflow-rs/crates/nomadflow-server/src/routes/features.rs
use std::sync::Arc;

use axum::{extract::State, http::StatusCode, routing::post, Json, Router};
use serde_json::{json, Value};

use nomadflow_core::models::{
    AttachBranchRequest, AttachBranchResponse, CreateFeatureRequest, CreateFeatureResponse,
    DeleteFeatureRequest, DeleteFeatureResponse, ListBranchesRequest, ListBranchesResponse,
    ListFeaturesRequest, ListFeaturesResponse, SwitchFeatureRequest, SwitchFeatureResponse,
};

use crate::state::AppState;

async fn list_features(
    State(state): State<Arc<AppState>>,
    Json(request): Json<ListFeaturesRequest>,
) -> Result<Json<ListFeaturesResponse>, (StatusCode, Json<Value>)> {
    match state.git.list_features(&request.repo_path).await {
        Ok(features) => Ok(Json(ListFeaturesResponse { features })),
        Err(e) => Err((
            StatusCode::INTERNAL_SERVER_ERROR,
            Json(json!({ "detail": e.to_string() })),
        )),
    }
}

async fn create_feature(
    State(state): State<Arc<AppState>>,
    Json(request): Json<CreateFeatureRequest>,
) -> Result<Json<CreateFeatureResponse>, (StatusCode, Json<Value>)> {
    let base_branch = if request.base_branch == "main" {
        None
    } else {
        Some(request.base_branch.as_str())
    };

    let (worktree_path, branch) = state
        .git
        .create_feature(&request.repo_path, &request.branch_name, base_branch)
        .await
        .map_err(|e| {
            (
                StatusCode::BAD_REQUEST,
                Json(json!({ "detail": e.to_string() })),
            )
        })?;

    // Inject hooks so Claude Code state tracking works in the new worktree
    let wt_path = std::path::Path::new(&worktree_path);
    if let Err(e) = state.agent_state.inject_hooks(wt_path).await {
        tracing::warn!(worktree = %worktree_path, "Failed to inject hooks on create: {e}");
    }

    let worktree_name = std::path::Path::new(&worktree_path)
        .file_name()
        .unwrap_or_default()
        .to_string_lossy()
        .to_string();

    Ok(Json(CreateFeatureResponse {
        worktree_path,
        branch,
        worktree_name,
    }))
}

async fn delete_feature(
    State(state): State<Arc<AppState>>,
    Json(request): Json<DeleteFeatureRequest>,
) -> Result<Json<DeleteFeatureResponse>, (StatusCode, Json<Value>)> {
    // Prevent deletion of main branch
    let features = state
        .git
        .list_features(&request.repo_path)
        .await
        .map_err(|e| {
            (
                StatusCode::INTERNAL_SERVER_ERROR,
                Json(json!({ "detail": e.to_string() })),
            )
        })?;

    if let Some(f) = features.iter().find(|f| f.name == request.feature_name) {
        if f.is_main {
            return Err((
                StatusCode::BAD_REQUEST,
                Json(json!({ "detail": "Cannot delete the main repository branch" })),
            ));
        }
    }

    // Resolve worktree path before deletion so we can clean up hooks and state files
    let repo_name = std::path::Path::new(&request.repo_path)
        .file_name()
        .unwrap_or_default()
        .to_string_lossy()
        .to_string();
    let worktree_path = state.settings.worktrees_dir().join(&repo_name).join(&request.feature_name);

    let deleted = state
        .git
        .delete_feature(&request.repo_path, &request.feature_name)
        .await
        .map_err(|e| {
            (
                StatusCode::INTERNAL_SERVER_ERROR,
                Json(json!({ "detail": e.to_string() })),
            )
        })?;

    // Clean up hooks and state file for the deleted worktree
    if deleted {
        if let Err(e) = state.agent_state.cleanup_hooks(&worktree_path).await {
            tracing::warn!(worktree = %worktree_path.display(), "Failed to cleanup hooks on delete: {e}");
        }
        let cwd = worktree_path.to_string_lossy().to_string();
        state.agent_state.delete_state_file(&cwd).await;
    }

    Ok(Json(DeleteFeatureResponse { deleted }))
}

async fn switch_feature(
    State(state): State<Arc<AppState>>,
    Json(request): Json<SwitchFeatureRequest>,
) -> Result<Json<SwitchFeatureResponse>, (StatusCode, Json<Value>)> {
    let features = state
        .git
        .list_features(&request.repo_path)
        .await
        .map_err(|e| {
            (
                StatusCode::INTERNAL_SERVER_ERROR,
                Json(json!({ "detail": e.to_string() })),
            )
        })?;

    let feature = features.iter().find(|f| f.name == request.feature_name);

    let worktree_path = if let Some(f) = feature {
        f.worktree_path.clone()
    } else {
        // Feature doesn't exist, create it
        let (wt, _branch) = state
            .git
            .create_feature(&request.repo_path, &request.feature_name, None)
            .await
            .map_err(|e| {
                (
                    StatusCode::BAD_REQUEST,
                    Json(json!({ "detail": e.to_string() })),
                )
            })?;
        wt
    };

    // Inject hooks into the worktree so Claude Code state tracking works
    let wt_path = std::path::Path::new(&worktree_path);
    if let Err(e) = state.agent_state.inject_hooks(wt_path).await {
        tracing::warn!(worktree = %worktree_path, "Failed to inject hooks on switch: {e}");
    }

    let worktree_name = std::path::Path::new(&worktree_path)
        .file_name()
        .unwrap_or_default()
        .to_string_lossy()
        .to_string();

    Ok(Json(SwitchFeatureResponse {
        switched: true,
        worktree_path,
        worktree_name,
    }))
}

async fn list_branches(
    State(state): State<Arc<AppState>>,
    Json(request): Json<ListBranchesRequest>,
) -> Result<Json<ListBranchesResponse>, (StatusCode, Json<Value>)> {
    let (branches, default_branch) =
        state
            .git
            .list_branches(&request.repo_path)
            .await
            .map_err(|e| {
                (
                    StatusCode::INTERNAL_SERVER_ERROR,
                    Json(json!({ "detail": e.to_string() })),
                )
            })?;

    Ok(Json(ListBranchesResponse {
        branches,
        default_branch,
    }))
}

async fn attach_branch(
    State(state): State<Arc<AppState>>,
    Json(request): Json<AttachBranchRequest>,
) -> Result<Json<AttachBranchResponse>, (StatusCode, Json<Value>)> {
    let (worktree_path, branch) = state
        .git
        .attach_branch(&request.repo_path, &request.branch_name)
        .await
        .map_err(|e| {
            (
                StatusCode::BAD_REQUEST,
                Json(json!({ "detail": e.to_string() })),
            )
        })?;

    // Inject hooks so Claude Code state tracking works in the attached worktree
    let wt_path = std::path::Path::new(&worktree_path);
    if let Err(e) = state.agent_state.inject_hooks(wt_path).await {
        tracing::warn!(worktree = %worktree_path, "Failed to inject hooks on attach: {e}");
    }

    let worktree_name = std::path::Path::new(&worktree_path)
        .file_name()
        .unwrap_or_default()
        .to_string_lossy()
        .to_string();

    Ok(Json(AttachBranchResponse {
        worktree_path,
        branch,
        worktree_name,
    }))
}

pub fn router() -> Router<Arc<AppState>> {
    Router::new()
        .route("/api/list-features", post(list_features))
        .route("/api/create-feature", post(create_feature))
        .route("/api/delete-feature", post(delete_feature))
        .route("/api/switch-feature", post(switch_feature))
        .route("/api/list-branches", post(list_branches))
        .route("/api/attach-branch", post(attach_branch))
}
